### PR TITLE
Parsing/Serialization for scroll-snap-points-*

### LINKED ISSUE #13871: Implement parsing/serialization for scroll-snap-points-x and scroll-snap-points-y
https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-snap-points-x

The grammar is `none | repeat(<length>)`. This means that it can either be `none`, or a function `repeat()` containing a length.

[`expect_function`](https://doc.servo.org/cssparser/parser/struct.Parser.html#method.expect_function) and [`parse_nested_block`](https://doc.servo.org/cssparser/parser/struct.Parser.html#method.parse_nested_block) will help with parsing the function. An example of this being used is [here](https://github.com/servo/servo/blob/4216224f9cc7f3430db2b59f4d77061824ba7a1e/components/style/properties/longhand/effects.mako.rs#L374), where it parses the `rect(...)` function.

You should make the `<%helpers:longhand>` block for `scroll-snap-points-y` just reexport the contents of `scroll-snap-points-x`

See https://github.com/servo/servo/wiki/Property-hacking-guide for instructions on how to work on such bugs. This bug requires familiarity with Rust.

File: https://github.com/servo/servo/blob/master/components/style/properties/longhand/box.mako.rs

### PR DESCRIPTION
**Pull request by ghost**
Implements Parsing/Serialization for scroll-snap-points-x and scroll-snap-points-y

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13871 

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14017)

### CODE REVIEW

**Comment by iamrohit7** [comment]
@Manishearth Please test out the changes in a Stylo build and let me know about any issues.

**Comment by Manishearth** [comment]
Please add single_keyword calls for scroll-snap-type-x and y.

**Comment by iamrohit7** [comment]
@Manishearth They are already in `box.mako.rs`

**Comment by Wafflespeanut** [comment]
We already have `LengthOrNone` for cases like this (which can be called with `predefined_type`), but we should `impl GeckoStyleCoordConvertible for LengthOrNone` in `gecko/values.rs` and add it to the list in `gecko.mako.rs` for auto-generation. So, maybe we should just land this, and refactor things later.

**Comment by Manishearth** [comment]
Your patch works, but Firefox actually supports percentage values too. We should change the `Length` to `LengthOrPercentage`, (and the `Au` to `values::computed::LengthOrPercentage`)

(We shouldn't use `LengthOrPercentageOrNone` because that parses differently)

I tried out the LengthOrPercentage version and that worked too, modified patch here:

<details>
```patch
diff --git a/servo/components/style/properties/gecko.mako.rs b/servo/components/style/properties/gecko.mako.rs
index 8edbafb..7c2aa42 100644
--- a/servo/components/style/properties/gecko.mako.rs
+++ b/servo/components/style/properties/gecko.mako.rs
@@ -850,7 +850,8 @@ fn static_assert() {
 </%self:impl_trait>
 
 <% skip_box_longhands= """display overflow-y vertical-align
-                          -moz-binding page-break-before page-break-after""" %>
+                          -moz-binding page-break-before page-break-after
+                          scroll-snap-points-x scroll-snap-points-y""" %>
 <%self:impl_trait style_struct_name="Box" skip_longhands="${skip_box_longhands}">
 
     // We manually-implement the |display| property until we get general
@@ -976,6 +977,24 @@ fn static_assert() {
 
     ${impl_simple_copy('page_break_after', 'mBreakAfter')}
 
+    pub fn set_scroll_snap_points_x(&mut self, v: longhands::scroll_snap_points_x::computed_value::T) {
+        match v.0 {
+            None => self.gecko.mScrollSnapPointsX.set_value(CoordDataValue::None),
+            Some(l) => l.to_gecko_style_coord(&mut self.gecko.mScrollSnapPointsX),
+        };
+    }
+
+    ${impl_coord_copy('scroll_snap_points_x', 'mScrollSnapPointsX')}
+
+    pub fn set_scroll_snap_points_y(&mut self, v: longhands::scroll_snap_points_y::computed_value::T) {
+        match v.0 {
+            None => self.gecko.mScrollSnapPointsY.set_value(CoordDataValue::None),
+            Some(l) => l.to_gecko_style_coord(&mut self.gecko.mScrollSnapPointsY),
+        };
+    }
+
+    ${impl_coord_copy('scroll_snap_points_y', 'mScrollSnapPointsY')}
+
 </%self:impl_trait>
 
 <%def name="simple_image_array_property(name, shorthand, field_name)">
diff --git a/servo/components/style/properties/longhand/box.mako.rs b/servo/components/style/properties/longhand/box.mako.rs
index 067c37e..8973f8b 100644
--- a/servo/components/style/properties/longhand/box.mako.rs
+++ b/servo/components/style/properties/longhand/box.mako.rs
@@ -841,6 +841,102 @@ ${helpers.keyword_list("animation-fill-mode",
     pub use super::transition_duration::SingleSpecifiedValue;
 </%helpers:longhand>
 
+<%helpers:longhand products="gecko" name="scroll-snap-points-y" animatable="False">
+    use cssparser::ToCss;
+    use std::fmt;
+    use values::LocalToCss;
+    use values::HasViewportPercentage;
+    use values::specified::LengthOrPercentage;
+
+    impl HasViewportPercentage for SpecifiedValue {
+        fn has_viewport_percentage(&self) -> bool {
+            match *self {
+                SpecifiedValue::Specified(length) => length.has_viewport_percentage(),
+                _ => false
+            }
+        }
+    }
+
+    pub mod computed_value {
+        use values::computed::LengthOrPercentage;
+
+        #[derive(Debug, Clone, PartialEq)]
+        #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+        pub struct T(pub Option<LengthOrPercentage>);
+    }
+
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub enum SpecifiedValue {
+        None,
+        Specified(LengthOrPercentage),
+    }
+
+    impl ToCss for computed_value::T {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            match self.0 {
+                None => dest.write_str("none"),
+                Some(l) => l.to_css(dest)
+            }
+        }
+    }
+    impl ToCss for SpecifiedValue {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            match *self {
+                SpecifiedValue::None => dest.write_str("none"),
+                SpecifiedValue::Specified(ref l) => l.to_css(dest),
+            }
+        }
+    }
+
+    #[inline]
+    pub fn get_initial_value() -> computed_value::T {
+        computed_value::T(None)
+    }
+
+    impl ToComputedValue for SpecifiedValue {
+        type ComputedValue = computed_value::T;
+
+        #[inline]
+        fn to_computed_value(&self, context: &Context) -> computed_value::T {
+            match *self {
+                SpecifiedValue::None => computed_value::T(None),
+                SpecifiedValue::Specified(l) =>
+                    computed_value::T(Some(l.to_computed_value(context))),
+            }
+        }
+        #[inline]
+        fn from_computed_value(computed: &computed_value::T) -> Self {
+            match *computed {
+                computed_value::T(None) => SpecifiedValue::None,
+                computed_value::T(Some(l)) =>
+                    SpecifiedValue::Specified(ToComputedValue::from_computed_value(&l))
+            }
+        }
+    }
+
+    pub fn parse(_context: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
+        if input.try(|input| input.expect_ident_matching("none")).is_ok() {
+            Ok(SpecifiedValue::None)
+        } else if input.try(|input| input.expect_function_matching("repeat")).is_ok() {
+            let i = input.parse_nested_block(|input| {
+                LengthOrPercentage::parse_non_negative(input).map(SpecifiedValue::Specified)
+            });
+            i
+        } else {
+            Err(())
+        }
+    }
+</%helpers:longhand>
+
+<%helpers:longhand products="gecko" name="scroll-snap-points-x" animatable="False">
+    pub use super::scroll_snap_points_y::SpecifiedValue;
+    pub use super::scroll_snap_points_y::computed_value;
+    pub use super::scroll_snap_points_y::get_initial_value;
+    pub use super::scroll_snap_points_y::parse;
+</%helpers:longhand>
+
+
 // CSSOM View Module
 // https://www.w3.org/TR/cssom-view-1/
 ${helpers.single_keyword("scroll-behavior",
</details>

File: components/style/properties/longhand/box.mako.rs [hunk 1]
```diff
@@ -841,6 +841,101 @@ ${helpers.keyword_list("animation-fill-mode",
...
+        #[derive(Debug, Clone, PartialEq)]
+        #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+        pub struct T(pub Option<Au>);
+    }
+
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub enum SpecifiedValue {
+        None,
+        Specified(Length),
```
Comment (Manishearth) [inline components/style/properties/longhand/box.mako.rs]: Probably could call it `Repeat`?

**Comment by Manishearth** [comment]
I don't want to use LengthOrNone because it parses differently.

**Comment by iamrohit7** [comment]
@Manishearth I applied the patch. Thanks! Should I also implement the scroll-snap-type as a shorthand? I'll have to find some time but I'll be able to do it in a couple of days.

**Comment by Manishearth** [comment]
@bors-servo r+

Yes, but try doing it in a separate PR (and add parse tests). Thanks!

**Comment by iamrohit7** [comment]
@Manishearth Sure. Thanks for the help!

**Comment by Manishearth** [comment]
Btw, the shorthand isn't specced, but implemented in Firefox (the spec details a more complicated longhand by the same name that deals with things Firefox doesn't implement). The shorthand is parsed as a _single_ keyword (not two), and sets both x/y keywords the same. The reverse conversion (from longhands to serialized shorthand) returns the empty string if both keywords are not the same.

**Comment by Manishearth** [comment]
@bors-servo retry

 - #13593

**Pull request merged**